src/core/solvers/ipopt/ipopt-iface.cpp

```diff
@@ -0,0 +1,665 @@
+      const std::size_t ndxi_next =
+          t + 1 == T_ ? problem_->get_terminalModel()->get_state()->get_ndx() : models[t + 1]->get_state()->get_ndx();
+      for (std::size_t idx_row = 0; idx_row < ndxi; idx_row++) {
+        for (std::size_t idx_col = 0; idx_col < (ndxi + nui + ndxi_next); idx_col++) {
+          iRow[idx_value] = ix + idx_row;
+          jCol[idx_value] = ixu + idx_col;
+          idx_value++;
+        }
+      }
+      ixu += ndxi + nui;
+      ix += ndxi;
+    }
+
+    // Initial condition
+    {
+      const std::size_t ndxi = problem_->get_runningModels()[0]->get_state()->get_ndx();
+      for (std::size_t idx_row = 0; idx_row < ndxi; idx_row++) {
+        for (std::size_t idx_col = 0; idx_col < ndxi; idx_col++) {
+          iRow[idx_value] = ix + idx_row;
+          jCol[idx_value] = idx_col;
+          idx_value++;
+        }
+      }
+    }
+
+    assert_pretty(nele_jac == idx_value,
+                  "Number of jacobian elements set does not coincide with the total non-zero Jacobian values");
+
+  } else {
+    std::size_t idx_value = 0;
+    std::size_t ixu = 0;
+    std::size_t ix = 0;
+    // Dynamic constraints
+    for (std::size_t t = 0; t < T_; t++) {
+      const boost::shared_ptr<ActionModelAbstract> &model = problem_->get_runningModels()[t];
+      const boost::shared_ptr<ActionDataAbstract> &data = problem_->get_runningDatas()[t];
+      const std::size_t ndxi = model->get_state()->get_ndx();
+      const std::size_t nui = model->get_nu();
+
+      const boost::shared_ptr<ActionModelAbstract> &model_next =
+          t + 1 == T_ ? problem_->get_terminalModel() : problem_->get_runningModels()[t + 1];
+      const boost::shared_ptr<ActionDataAbstract> &data_next =
+          t + 1 == T_ ? problem_->get_terminalData() : problem_->get_runningDatas()[t + 1];
+      const std::size_t ndxi_next = model_next->get_state()->get_ndx();
+
+      datas_[t]->dx = Eigen::VectorXd::Map(x + ixu, ndxi);
+      datas_[t]->u = Eigen::VectorXd::Map(x + ixu + ndxi, nui);
+      datas_[t + 1]->dx = Eigen::VectorXd::Map(x + ixu + ndxi + nui, ndxi_next);
+
+      model->get_state()->integrate(xs_[t], datas_[t]->dx, datas_[t]->x);
+      model_next->get_state()->integrate(xs_[t + 1], datas_[t + 1]->dx, datas_[t + 1]->x);
+
+      model->calcDiff(data, datas_[t]->x, datas_[t]->u);
+
+      model_next->get_state()->Jintegrate(xs_[t + 1], datas_[t + 1]->dx, datas_[t + 1]->Jsum_dx,
+                                          datas_[t + 1]->Jsum_dx, second, setto);  // datas_[t]->Jsum_dxnext == eq. 81
+      model->get_state()->Jdiff(data->xnext, datas_[t + 1]->x, datas_[t]->Jdiff_x, datas_[t + 1]->Jdiff_x,
+                                both);  // datas_[t+1]->Jdiff_x == eq. 83, datas_[t]->Jdiff_x == eq.82
+      model->get_state()->Jintegrate(xs_[t], datas_[t]->dx, datas_[t]->Jsum_dx, datas_[t]->Jsum_dx, second,
+                                     setto);  // datas_[t]->Jsum_dx == eq. 81
```

Comment:
If we create data for next values, then we could parallelise this routine.
Additionally, there are other routines that are possible to parallelise now.

Reply:
Optimization related